govcd/catalog.go

```diff
@@ -24,18 +24,94 @@ import (
+type CatalogOperations interface {
+	FindCatalogItem(catalogitem string) (CatalogItem, error)
+}
+
+// AdminCatalog is a admin view of a vCloud Director Catalog
+// To be able to get an AdminCatalog representation, users must have
+// admin credentials to the System org. AdminCatalog is used
+// for creating, updating, and deleting a Catalog.
+type AdminCatalog struct {
+	AdminCatalog *types.AdminCatalog
+	c            *Client
+}
+
 type Catalog struct {
 	Catalog *types.Catalog
 	c       *Client
 }
 
-func NewCatalog(c *Client) *Catalog {
+func NewCatalog(client *Client) *Catalog {
 	return &Catalog{
 		Catalog: new(types.Catalog),
-		c:       c,
+		c:       client,
+	}
+}
+
+func NewAdminCatalog(client *Client) *AdminCatalog {
+	return &AdminCatalog{
+		AdminCatalog: new(types.AdminCatalog),
+		c:            client,
 	}
 }
 
+// Deletes the Catalog, returning an error if the vCD call fails.
+func (adminCatalog *AdminCatalog) Delete(force, recursive bool) error {
+	adminCatalogHREF := adminCatalog.c.VCDHREF
+	adminCatalogHREF.Path += "/admin/catalog/" + adminCatalog.AdminCatalog.ID[19:]
+
+	req := adminCatalog.c.NewRequest(map[string]string{
+		"force":     strconv.FormatBool(force),
+		"recursive": strconv.FormatBool(recursive),
+	}, "DELETE", adminCatalogHREF, nil)
+
+	_, err := checkResp(adminCatalog.c.Http.Do(req))
+
+	if err != nil {
+		return fmt.Errorf("error deleting Catalog %s: %s", adminCatalog.AdminCatalog.ID, err)
+	}
+
+	return nil
+}
+
+//   Updates the Catalog definition from current Catalog struct contents.
+//   Any differences that may be legally applied will be updated.
+//   Returns an error if the call to vCD fails.
+func (adminCatalog *AdminCatalog) Update() (Task, error) {
+	vcomp := &types.AdminCatalog{
+		Xmlns:       "http://www.vmware.com/vcloud/v1.5",
+		Name:        adminCatalog.AdminCatalog.Name,
+		Description: adminCatalog.AdminCatalog.Description,
+		IsPublished: adminCatalog.AdminCatalog.IsPublished,
+	}
+	adminCatalogHREF, err := url.ParseRequestURI(adminCatalog.AdminCatalog.HREF)
+	if err != nil {
+		return Task{}, fmt.Errorf("error parsing admin catalog's href: %v", err)
+	}
+	output, err := xml.MarshalIndent(vcomp, "  ", "    ")
+	if err != nil {
+		return Task{}, fmt.Errorf("error marshalling xml data for update", err)
+	}
+	xmlData := bytes.NewBufferString(xml.Header + string(output))
+	req := adminCatalog.c.NewRequest(map[string]string{}, "PUT", *adminCatalogHREF, xmlData)
+	req.Header.Add("Content-Type", "application/vnd.vmware.admin.catalog+xml")
+	resp, err := checkResp(adminCatalog.c.Http.Do(req))
+	if err != nil {
+		return Task{}, fmt.Errorf("error updating catalog: %s : %s", err, adminCatalogHREF.Path)
+	}
+
+	catalog := NewAdminCatalog(adminCatalog.c)
+	if err = decodeBody(resp, catalog.AdminCatalog); err != nil {
+		return Task{}, fmt.Errorf("error decoding task response: %s", err)
+	}
+	adminCatalog.AdminCatalog = catalog.AdminCatalog
+	return Task{
+		Task: catalog.AdminCatalog.Tasks.Task[0],
```

Comment:
This operation can easily fail. We are assuming that there is a `catalog.AdminCatalog.Tasks.Task[0]` element without checking whether `catalog.AdminCatalog.Tasks` is defined. In fact, the test fails exactly at this point with an invalid pointer dereference error.

Thus, what does it mean in this case to have an operation that responds with "200 OK" but there is no task? Is the task expected?
Please add the URL of the API being used.